metronome/networking/src/io/iohk/metronome/networking/ConnectionHandler.scala

```diff
@@ -0,0 +1,274 @@
+  ConnectionAlreadyClosed,
+  ConnectionError
+}
+
+import java.net.InetSocketAddress
+import scala.util.control.NoStackTrace
+
+class ConnectionHandler[F[_]: Concurrent, K, M](
+    connectionQueue: ConcurrentQueue[F, HandledConnection[F, K, M]],
+    connectionsRegister: ConnectionsRegister[F, K, M],
+    messageQueue: ConcurrentQueue[F, MessageReceived[K, M]],
+    cancelToken: TryableDeferred[F, Unit],
+    connectionFinishCallback: HandledConnection[F, K, M] => F[Unit]
+) {
+
+  private val numberOfRunningConnections = AtomicInt(0)
+
+  private def closeAndDeregisterConnection(
+      handledConnection: HandledConnection[F, K, M]
+  ): F[Unit] = {
+    for {
+      _ <- Concurrent[F].delay(numberOfRunningConnections.decrement())
+      _ <- connectionsRegister.deregisterConnection(handledConnection)
+      _ <- handledConnection.close
+    } yield ()
+  }
+
+  /** Registers connections and start handling incoming messages in background, in case connection is already handled
+    * it closes it
+    *
+    * @param possibleNewConnection, possible connection to handle
+    */
+  def registerOrClose(
+      possibleNewConnection: HandledConnection[F, K, M]
+  ): F[Unit] = {
+    connectionsRegister.registerIfAbsent(possibleNewConnection).flatMap {
+      case Some(_) =>
+        //TODO [PM-3092] for now we are closing any new connections in case of conflict, we may investigate other strategies
+        // like keeping old for outgoing and replacing for incoming
+        possibleNewConnection.close
+      case None =>
+        connectionQueue.offer(possibleNewConnection)
+    }
+  }
+
+  /** Checks if handler already handles connection o peer with provided key
+    *
+    * @param connectionKey key of remote peer
+    */
+  def isNewConnection(connectionKey: K): F[Boolean] = {
+    connectionsRegister.isNewConnection(connectionKey)
+  }
+
+  /** Retrieves set of keys of all connected and handled peers
+    */
+  def getAllActiveConnections: F[Set[K]] =
+    connectionsRegister.getAllRegisteredConnections.map { connections =>
+      connections.map(_.key)
+    }
+
+  /** Number of connections actively red in background
+    */
+  def numberOfActiveConnections: F[Int] = {
+    Concurrent[F].delay(numberOfRunningConnections.get())
+  }
+
+  /** Stream of all messages received from all remote peers
+    */
+  def incomingMessages: Iterant[F, MessageReceived[K, M]] =
+    Iterant.repeatEvalF(messageQueue.poll)
+
+  /** Retrieves handled connection if one exists
+    *
+    * @param key, key of remote peer
+    */
+  def getConnection(key: K): F[Option[HandledConnection[F, K, M]]] =
+    connectionsRegister.getConnection(key)
+
+  def sendMessage(
+      recipient: K,
+      message: M
+  ): F[Either[ConnectionAlreadyClosedException[K], Unit]] = {
+    getConnection(recipient).flatMap {
+      case Some(connection) =>
+        connection
+          .sendMessage(message)
+          .attemptNarrow[ConnectionAlreadyClosed]
```

Comment:
What does this do exactly?

Reply:
It is similar to attempt i.e error materialisation, but it handles only subset of errors. So we can only recover types of errors which are interesting to us, but leave out some unexpected one.

Reply:
Right, so the `case Left(_)` below will only be called if the error is [already an instance of](https://github.com/typelevel/cats/blob/main/core/src/main/scala/cats/ApplicativeError.scala#L84) `ConnectionAlreadyClosed`, and I see that the Scalanet provider [raises this exception](https://github.com/input-output-hk/metronome/pull/7/files/97c3f272e48681e25766bee4612c4da3d3651ce7#diff-cdd6bc91c414dfa9f49a5902222bfb6eecd67691ef8f344cce579beed02c382fR48), which is good, I haven't noticed before.

So maybe it should be `closed: Left => connection.close.as(closed)`?